### ISSUE TITLE
Could NOT find OASIS3MCT

### ISSUE THREAD

**Issue by mvhulten** [HPSCTerrSys/eTSMP#4]
I tried to build eCLM-ParFlow on juwels, following the steps as described in the [README](../blob/master/README.md).

CMake failed:

```
[vanhulten1@jwlogin07 eTSMP]$ cmake -S . -B ${BUILD_DIR}                  \
>       -DCMAKE_INSTALL_PREFIX="$INSTALL_DIR" \
>       -DeCLM_SRC=${eCLM_SRC}                \                                                                                                                              
>       -DPARFLOW_SRC=${PARFLOW_SRC}     
                                                                                                                                                                             
-- The C compiler identification is Intel 2021.6.0.20220226
-- The CXX compiler identification is Intel 2021.6.0.20220226
-- The Fortran compiler identification is Intel 2021.6.0.20220226
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /p/software/juwels/stages/2022/software/psmpi/5.5.0-1-intel-compilers-2021.4.0/bin/mpicc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /p/software/juwels/stages/2022/software/psmpi/5.5.0-1-intel-compilers-2021.4.0/bin/mpicxx - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /p/software/juwels/stages/2022/software/psmpi/5.5.0-1-intel-compilers-2021.4.0/bin/mpif90 - skipped
CMake Error at /p/software/juwels/stages/2023/software/CMake/3.23.1-GCCcore-11.3.0/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OASIS3MCT (missing: PSMILE_LIB MCT_LIB MPEU_LIB SCRIP_LIB)
Call Stack (most recent call first):
  /p/software/juwels/stages/2023/software/CMake/3.23.1-GCCcore-11.3.0/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindOASIS3MCT.cmake:8 (find_package_handle_standard_args)
  CMakeLists.txt:13 (find_package)


-- Configuring incomplete, errors occurred!
See also "/p/home/jusers/vanhulten1/juwels/software/repos/eTSMP/bld/JUWELS_eCLM-ParFlow/CMakeFiles/CMakeOutput.log".
```

**Comment by mvhulten** [comment]
It seems that OASIS is missing.  In step 4, should there not be also the OASIS "component"?  Or a mention of requiring OASIS (albeit maybe trivial for some)?

I pulled in the eCLM and ParFlow components.  Only below `./eCLM/src/clm5/oasis3/` I found some OASIS-related source code files.

I do not have access to https://icg4geo.icg.kfa-juelich.de/ , which is why I chose for now to compile eCLM-ParFlow: this requires only accessible repositories.

**Comment by mvhulten** [comment]
I got access to OASIS.

For now, this is only a documentation issue.

**Comment by kvrigor** [comment]
Hi! CMake should automatically find OASIS after running `source env/jsc.2022_Intel.sh`. This script automatically sets the path to the shared OASIS library:

https://github.com/HPSCTerrSys/eTSMP/blob/5a8c08c69d735f54ff831a8ebb60c3dc14edcafd/env/jsc.2022_Intel.sh#L30-L34

One problem here is you may need access to the `cslts` compute project. Do you have access to `cslts` or will you be using another compute project for your runs?

**Comment by mvhulten** [comment]
Thank you for your help!

I do not have access to `cslts`.  Maybe I should ask for access.
I am in the projects `cdetect`, `detectdata` and `moistly`.

I have compiled <https://gitlab.com/cerfacs/oasis3-mct>, but it is unclear to me how to use it / what to set `OASIS_ROOT` to.
(I tried a path containing a `lib` directory but not `include`, but this is all a bit off-topic.)

**Comment by kvrigor** [comment]
After discussing with @s-poll and @chartick a few days ago, I realized that letting eTSMP reuse a built `OASIS3-MCT` library causes more headache than I initially thought since not all users have access to the `cslts` shared folder. Even if the libraries were hosted on a "more public" folder (such as what @DCaviedesV suggests), the overhead of maintaining the libraries becomes another issue. To simplify things for both users and maintainers, I am changing eTSMP's default behavior into building `OASIS3-MCT`, which is the same with what TSMP is doing.

@mvhulten it would be nice if you can test this new feature which hopefully could solve your issue. You will have to switch to [dev-build-oasis](https://github.com/HPSCTerrSys/eTSMP/tree/dev-build-oasis) branch to try this. Below's a recipe that you can follow:

```bash
# 1. Download OASIS3-MCT
git clone https://icg4geo.icg.kfa-juelich.de/ExternalReposPublic/oasis3-mct
OASIS_SRC=`realpath oasis3-mct`

# 2. Use eTSMP `dev-build-oasis` branch
cd /path/to/your/eTSMP
git pull && git checkout dev-build-oasis

# 3. Load environment variables
source env/jsc.2023_Intel.sh

# 4. Download component models that you may need. (See step 4 from 
#    the README: https://github.com/HPSCTerrSys/eTSMP#quickstart )

# 5. Sample CMake incantation for eCLM-ParFlow. Note that CMake now needs the 
#    source directory for OASIS3-MCT so that it could build it.
cmake -S . -B ${BUILD_DIR}                  \
      -DCMAKE_INSTALL_PREFIX="$INSTALL_DIR" \
      -DOASIS_SRC=${OASIS_SRC}              \
      -DeCLM_SRC=${eCLM_SRC}                \
      -DPARFLOW_SRC=${PARFLOW_SRC}

cmake --build ${BUILD_DIR}
cmake --install ${BUILD_DIR}
```

**Comment by mvhulten** [comment]
Thanks, I've tried this (including step 3 of quick start as well, of course).
This happens:

```
CMake Error at /p/software/juwels/stages/2023/software/CMake/3.23.1-GCCcore-11.3.0/share/cmake-3.23/Modules/ExternalProject.cmake:2776 (message):
  No download info given for 'eCLM' and its source directory:

   /p/home/jusers/vanhulten1/juwels/software/repos/eTSMP/bld/JUWELS_eCLM-ParFlow/eCLM/src/eCLM

  is not an existing non-empty directory.  Please specify one of:
```

[CMakeOutput.log](https://github.com/HPSCTerrSys/eTSMP/files/12417634/CMakeOutput.log)

I have not yet looked closer at it; will do later today.

**Comment by mvhulten** [comment]
`bld/JUWELS_eCLM-ParFlow/eCLM/src/eCLM/` is indeed empty.

Should this be populated in a previous step?

Apropos, `eCLM/src/` is not empty (and contains a subdirectory `eclm`—note the case).

Above paths are w.r.t. root of eTSMP repo.

**Comment by mvhulten** [comment]
@kvrigor, could I get priviledges to push a branch?

**Comment by kvrigor** [comment]
@mvhulten Kindly `git pull` the latest changes and try to build again.

> @kvrigor, could I get privileges to push a branch?

I believe only @HPSC-TerrSys members are allowed to do such (perhaps @DCaviedesV could also comment on this). But feel free to fork and raise a PR for any bugfixes and/or improvements 😃

**Comment by mvhulten** [comment]
I was able to build the eCLM-ParFlow combination!

This morning, I tried by accident a build of eCLM-ICON (which failed, but that is not important now).  Then I just went on with eCLM-ParFlow, but it still failed because it still tried to include ICON.  Is that normal behaviour?  Is there a way to run a `make clean` (through CMake)?

In the end, I built eCLM-ParFlow simply by starting with a clean checkout of eTSMP.
